File: examples/cfg.cpp
#include <a0.h>
#include <iostream>
#include <nlohmann/json.hpp>

int main() {
  setvbuf(stdout, NULL, _IONBF, 0);

  a0::Cfg cfg("state");
  
  cfg.write(nlohmann::json{
      {"a", 1},
      {"b", {{"c", "Hello, World!"}}},
  });
  
  auto a = cfg.var<int>("/a");
  auto full_cfg = cfg.var<nlohmann::json>();

  std::cout << "original:\n  a=" << *a << std::endl;
  
  cfg.mergepatch({{"a", 2}});
  
  std::cout << "changed, but not updated:\n  a="
            << *a << std::endl;

  cfg.update_var();
  
  std::cout << "updated:\n  a="
            << *a << std::endl;
            
  std::cout << "\nfull cfg:\n"
            << full_cfg->dump(2) << std::endl;
}

File: examples/cfg.py
import a0
import json

a0.Cfg("state").write(
    json.dumps(
        {
            "a": 1,
            "b": {"c": "Hello, World!"},
        }
    )
)


class Msg:
    def __init__(self, c):
        self.msg = c


whole = a0.cfg("state")
print(whole)

a = a0.cfg("state", "/a")
print(a)

ai = a0.cfg("state", "/a", int)
print(ai)

c = a0.cfg("state", "/b/c")
print(c)

cs = a0.cfg("state", "/b/c", Msg)
print(cs.msg)


a0.Cfg("state").write(
    json.dumps(
        {
            "a": 2,
            "b": {"c": "Goodbye, World!"},
        }
    )
)

print(cs.msg)

a0.update_configs()

print(cs.msg)
